### PR TITLE
feat(ldap): add option to load ldap from file

### PR DESCRIPTION
**Pull request by laurentiuNiculae**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/config/config.go [hunk 1]
```diff
@@ -102,6 +102,7 @@ type SchedulerConfig struct {
 }
 
 type LDAPConfig struct {
+	LDAPConfigFile     string
```
Comment (rchincha) [inline pkg/api/config/config.go:105]: LDAPCredentialsFile only pls, no need to put all config in that file.

Comment (andaaron) [reply pkg/api/config/config.go:105]: To clarify, credentials being "BindDN" and "BindPassword".

Comment (rchincha) [reply pkg/api/config/config.go:105]: yup, and that can be in JSON

File: pkg/api/config/config.go [hunk 2]
```diff
@@ -102,18 +102,19 @@ type SchedulerConfig struct {
...
-	SkipVerify         bool
-	SubtreeSearch      bool
-	Address            string
-	BindDN             string
-	UserGroupAttribute string
-	BindPassword       string
-	BaseDN             string
-	UserAttribute      string
-	CACert             string
+	LDAPCredentialsFile string
```
Comment (andaaron) [inline pkg/api/config/config.go:105]:
```suggestion
	CredentialsFile string
```

File: pkg/cli/root.go [hunk 3]
```diff
@@ -775,6 +780,47 @@ func LoadConfiguration(config *config.Config, configPath string) error {
...
+		return err
+	}
+
+	conf.HTTP.Auth.LDAP.BindDN = newLDAPCredentials.BindDN
+	conf.HTTP.Auth.LDAP.BindPassword = newLDAPCredentials.BindPassword
+
+	return nil
+}
+
+type LdapCredentials struct {
```
Comment (andaaron) [inline pkg/cli/root.go:803]: Maybe this should be defined in config.go.

File: pkg/api/config/config.go [hunk 4]
```diff
@@ -101,7 +101,13 @@ type SchedulerConfig struct {
...
 }
 
+type LDAPCredentials struct {
+	BindDN       string
+	BindPassword string
+}
+
 type LDAPConfig struct {
+	CredentialsFile    string
 	Port               int
```
Comment (rchincha) [inline pkg/api/config/config.go:111]: BindPassword string below should be removed right? In fact, isn't that the only credential?

File: pkg/cli/root.go [hunk 5]
```diff
@@ -771,6 +776,42 @@ func LoadConfiguration(config *config.Config, configPath string) error {
 	return nil
 }
 
+func updateLDAPConfig(conf *config.Config) error {
+	if conf.HTTP.Auth == nil || conf.HTTP.Auth.LDAP == nil {
+		return nil
+	}
+
+	if conf.HTTP.Auth.LDAP.CredentialsFile == "" {
```
Comment (rchincha) [inline pkg/cli/root.go:784]: This credentials file may be empty (or not specified in config)
But the rest is still valid. We could have anonymous LDAP binds?

File: pkg/api/config/config.go (hunk 4 above)
Comment (andaaron) [reply pkg/api/config/config.go:111]: That is the password, the username is the other property.

And why remove it from here? I thought this feature was optional?

File: pkg/cli/root.go (hunk 5 above)
Comment (andaaron) [reply pkg/cli/root.go:784]: Why would the current implementation need changes? It doesn't make an update to the credentials in the config if the file is not specified?

File: pkg/api/config/config.go (hunk 4 above)
Comment (andaaron) [reply pkg/api/config/config.go:111]: Optional in the sense the user may choose to keep his credentials in the config, or he may choose to save them in the credentials file.

Comment (rchincha) [reply pkg/api/config/config.go:111]: No, we always want separate credentials file

File: pkg/cli/root.go (hunk 5 above)
Comment (rchincha) [reply pkg/cli/root.go:784]: Similar to sync credentials or all other credentials

Comment (rchincha) [reply pkg/cli/root.go]: CredentialsFile can be nil and add a test case to test LDAP code works after that.

File: pkg/api/config/config.go [hunk 6]
```diff
@@ -121,16 +121,23 @@ type SchedulerConfig struct {
...
+	CredentialsFile    string
+	Unauthenticated    bool
 	Port               int
 	Insecure           bool
 	StartTLS           bool // if !Insecure, then StartTLS or LDAPs
 	SkipVerify         bool
 	SubtreeSearch      bool
 	Address            string
-	BindDN             string
+	BindDN             string `json:"-"`
```
Comment (rchincha) [inline pkg/api/config/config.go:138]: Remove these completely from here.

File: pkg/api/config/config.go [hunk 7]
```diff
@@ -121,16 +121,23 @@ type SchedulerConfig struct {
 	NumWorkers int
 }
 
+type LDAPCredentials struct {
+	BindDN       string
+	BindPassword string
+}
+
 type LDAPConfig struct {
+	CredentialsFile    string
```
Comment (rchincha) [inline pkg/api/config/config.go:130]: This can be "" so expect that and test for it.

File: pkg/api/config/config.go (hunk 6 above)
Comment (laurentiuNiculae) [reply pkg/api/config/config.go:138]: We can't remove them. We need a way to access their value from the config. How can we transmit their values through the config if we don't set them here?

Comment (rchincha) [reply pkg/api/config/config.go:138]: Can it become lowercase so limited in scope (or) put it behind Getter methods.

**Comment by andaaron** [comment]
Can you also add an example configuration file?

File: pkg/api/ldap.go [hunk 8]
```diff
@@ -215,11 +232,20 @@ func (lc *LDAPClient) Authenticate(username, password string) (bool, map[string]
...
-		lc.Log.Error().Err(err).Str("bindDN", userDN).Msg("user bind failed")
+	if password != "" {
+		err = lc.Conn.Bind(userDN, password)
+		if err != nil {
+			lc.Log.Error().Err(err).Str("bindDN", userDN).Msg("user bind failed")
 
-		return false, user, userGroups, err
+			return false, user, userGroups, err
+		}
+	} else {
```
Comment (andaaron) [inline pkg/api/ldap.go:242]: What use case are we supporting here? All users we are authenticating should have valid passwords?

File: pkg/api/ldap.go [hunk 9]
```diff
@@ -127,7 +128,7 @@ func (lc *LDAPClient) Authenticate(username, password string) (bool, map[string]
 	lc.lock.Lock()
 	defer lc.lock.Unlock()
 
-	if password == "" {
+	if password == "" && !lc.Unauthenticated {
```
Comment (andaaron) [inline pkg/api/ldap.go:131]: Why do we need this change?

File: pkg/api/ldap.go
Comment (andaaron) [inline pkg/api/ldap.go:1]: I don't understand the changes in this file.

The 2 cases I would expect are:
A) The config can have a bind user and password mentioned in the credentials file. In that case we should use it to search for the user we intend to authenticate with. If that user exists, we should bind with his username and his password to verify the password as well.

B) If we don't have a credentials file, so no bind user or password (there may be a case for bind user without a bind password, do we need to support it?), we can do an anonymous bind to search for the user, and then we should bind with his username and his password to verify the password.

Maybe I misunderstand the changes, but why would we fail if the bind username or password is not specified? And in a separate case why should we allow authentication of LDAP users without passwords?

File: pkg/cli/server/root.go [hunk 10]
```diff
@@ -752,6 +756,54 @@ func LoadConfiguration(config *config.Config, configPath string) error {
 	return nil
 }
 
+func updateLDAPConfig(conf *config.Config) error {
+	if conf.HTTP.Auth == nil || conf.HTTP.Auth.LDAP == nil {
+		return nil
+	}
+
+	if conf.HTTP.Auth.LDAP.Unauthenticated {
+		conf.HTTP.Auth.LDAP.SetBindDN("anonym-user")
```
Comment (andaaron) [inline pkg/cli/server/root.go:765]: Why do we need this?

File: pkg/cli/server/root.go [hunk 11]
```diff
@@ -752,6 +756,54 @@ func LoadConfiguration(config *config.Config, configPath string) error {
...
+
+		return nil
+	}
+
+	if conf.HTTP.Auth.LDAP.CredentialsFile == "" {
```
Comment (andaaron) [inline pkg/cli/server/root.go:770]: If the credentials file is empty we should assume we need to do an anonymous bind to search for users instead?

File: pkg/api/controller_test.go [hunk 12]
```diff
@@ -2077,6 +2099,332 @@ func TestBasicAuthWithLDAP(t *testing.T) {
...
+			"Auth": {
+				"LDAP": {
+					"CredentialsFile":    "%s",
+					"BaseDN":             "%v",
+					"UserAttribute":      "%v",
+					"UserGroupAttribute": "memberOf",
+					"Insecure":           true,
+					"Address":            "%v",
+					"Port":               %v,
+					"BindDN":             "%v",
```
Comment (andaaron) [inline pkg/api/controller_test.go:2342]:
```suggestion
```

File: pkg/api/controller_test.go [hunk 13]
```diff
@@ -2077,6 +2099,332 @@ func TestBasicAuthWithLDAP(t *testing.T) {
...
+					"BindPassword":       "%v"
```
Comment (andaaron) [inline pkg/api/controller_test.go:2343]:
```suggestion
```

File: pkg/cli/server/root_test.go [hunk 14]
```diff
@@ -1447,6 +1448,134 @@ func TestScrub(t *testing.T) {
...
+		server.SetArgs([]string{"serve", configPath})
+		So(func() { err = server.Execute() }, ShouldPanic)
+	})
+
+	Convey("unauthenticated LDAP config", t, func() {
+		tempDir := t.TempDir()
+		ldapConfigContent := "{}"
+		ldapConfigPath := filepath.Join(tempDir, "ldap.json")
+
+		err := os.WriteFile(ldapConfigPath, []byte(ldapConfigContent), 0o000)
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/cli/server/root_test.go]: We don't need it anymore

**Review by andaaron — APPROVED** [review]

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**